backend/press/press.go

```diff
@@ -0,0 +1,1376 @@
+			return
+		}
+		notifyFunc(wrappedPath, entryType)
+	}
+	do(ctx, wrappedNotifyFunc, pollIntervalChan)
+}
+
+// PublicLink generates a public link to the remote path (usually readable by anyone)
+func (f *Fs) PublicLink(ctx context.Context, remote string, duration fs.Duration, unlink bool) (string, error) {
+	do := f.Fs.Features().PublicLink
+	if do == nil {
+		return "", errors.New("PublicLink not supported")
+	}
+	o, err := f.NewObject(ctx, remote)
+	if err != nil {
+		// assume it is a directory
+		return do(ctx, remote, duration, unlink)
+	}
+	return do(ctx, o.(*Object).Object.Remote(), duration, unlink)
+}
+
+/*** OBJECT FUNCTIONS ***/
+
+// ObjectMetadata describes the metadata for an Object.
+type ObjectMetadata struct {
+	Size                int64  // Uncompressed size of the file.
+	Mode                int    // Compression mode of the file.
+	Hash                []byte // MD5 hash of the file.
+	MimeType            string // Mime type of the file
+	CompressionMetadata sgzip.GzipMetadata
+}
+
+// Object with external metadata
+type Object struct {
+	fs.Object                 // Wraps around data object for this object
+	f         *Fs             // Filesystem object is in
+	mo        fs.Object       // Metadata object for this object
+	moName    string          // Metadata file name for this object
+	size      int64           // Size of this object
+	meta      *ObjectMetadata // Metadata struct for this object (nil if not loaded)
+}
+
+// This function generates a metadata object
+func newMetadata(size int64, mode int, cmeta sgzip.GzipMetadata, hash []byte, mimeType string) *ObjectMetadata {
+	meta := new(ObjectMetadata)
+	meta.Size = size
+	meta.Mode = mode
+	meta.CompressionMetadata = cmeta
+	meta.Hash = hash
+	meta.MimeType = mimeType
+	return meta
+}
+
+// This function will read the metadata from a metadata object.
+func readMetadata(ctx context.Context, mo fs.Object) (meta *ObjectMetadata) {
+	// Open our meradata object
+	rc, err := mo.Open(ctx)
+	if err != nil {
+		return nil
+	}
+	defer func() {
+		err := rc.Close()
+		if err != nil {
+			fs.Errorf(mo, "Error closing object: %v", err)
+		}
+	}()
+	// Read gzipped compressed data from it
+	gzipReader, err := gzip.NewReader(rc)
+	if err != nil {
+		return nil
+	}
+	// Decode the gob from that
+	meta = new(ObjectMetadata)
+	metadataDecoder := gob.NewDecoder(gzipReader)
+	err = metadataDecoder.Decode(meta)
+	// Cleanup and return
+	_ = gzipReader.Close() // We don't really care if this is closed properly
+	if err != nil {
+		return nil
+	}
+	return meta
+}
+
+// Remove removes this object
+func (o *Object) Remove(ctx context.Context) error {
+	err := o.loadMetadataObjectIfNotLoaded(ctx)
+	if err != nil {
+		return err
+	}
+	err = o.mo.Remove(ctx)
+	objErr := o.Object.Remove(ctx)
+	if err != nil {
+		return err
+	}
+	return objErr
+}
+
+// ReadCloserWrapper combines a Reader and a Closer to a ReadCloser
+type ReadCloserWrapper struct {
+	dataSource io.Reader
```

Comment:
If you embed both of these

```go
type ReadCloserWrapper struct {
    io.Reader
    io.Closer
}
```

Then you don't need to implement the `Read` and `Close` methods.

You initialise with `ReadCloserWrapper{Reader: r, Closer: c}`

We implement this at least once elsewhere!

https://github.com/rclone/rclone/blob/d8239e01941b69fcdef2a62fad8791c393a058ec/fs/operations/operations.go#L1007-L1011

So maybe it should go into `lib/readers`, however it is a very small adaptor so I'm not sure it is worth it.

Reply:
I think thats to small to but put into `lib/readers`